File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:

    int numIslands(vector<vector<char>>& grid) 
    {
        
        if(grid.empty())
            return 0;

        int rows = grid.size();
        int cols = grid[0].size();
        int count = 0;

        //traverse, check if it's 1

        for(int i = 0; i < rows; i++)
        {
            for(int j = 0; j < cols; j++)
            {
                if(grid[i][j] == '1')
                {
                    // grid[i][j] = '0';
                    // cout << grid[i][j];
                    dfs(i,j,grid);
                    count++;
                }
            }
        }
        return count;

    }

    void dfs(int i, int j, vector<vector<char>>& g)
    {
        int r = g.size();
        int c = g[0].size();

        bool legit;

        if(i>=0 && i<r && j>=0 && j<c)
            legit = true;
        else
            legit = false;

        if(legit == true && g[i][j] == '1')
        {
            g[i][j] = '0';
            dfs(i,j-1,g);
            dfs(i,j+1,g);
            dfs(i-1,j,g);
            dfs(i+1,j,g);
        }
        else
            return;

    }

};

File: Data Structures & Algorithms/valid-sudoku/submission-0.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        rows = collections.defaultdict(set)
        cols = collections.defaultdict(set)
        squares = collections.defaultdict(set)

        for r in range(9):
            for c in range(9):
                value = board[r][c]
                # check for empty elements and continue 
                if value == ".":
                    continue

                # check if duplicate value in that row / in that column / in that square
                if( value in rows[r] or value in cols[c] or value in squares[(r // 3, c // 3)] ):
                    return False 

                # add into set if new value
                cols[c].add(value)
                rows[r].add(value)
                squares[(r // 3,c // 3)].add(value)

        return True
        
        